File: Sources/SanitizerKit/Apply.swift
import Foundation

// Executes selected changes against the library and writes a reversal log so
// any run can be undone. Same log shape/location as the Ruby apply.rb, so the
// two are interoperable. Ported from apply.rb.
struct Apply {
    var client: Client
    var logDir: URL

    init(client: Client = Client(), logDir: URL = Config.home.appendingPathComponent("logs")) {
        self.client = client
        self.logDir = logDir
        try? FileManager.default.createDirectory(at: logDir, withIntermediateDirectories: true)
    }

    @discardableResult
    func run(removeIDs: [String], addIDs: [String]) async throws -> URL {
        if !removeIDs.isEmpty { try await client.delete("/me/tracks", ids: removeIDs) }
        if !addIDs.isEmpty { try await client.put("/me/tracks", ids: addIDs) }
        return try writeLog(removed: removeIDs, added: addIDs)
    }

    // Inverts a reversal log: re-like what we removed, unlike what we added.
    func undo(logURL: URL) async throws -> (reliked: Int, unliked: Int) {
        let obj = (try? JSONSerialization.jsonObject(with: Data(contentsOf: logURL))) as? [String: Any] ?? [:]
        let removed = obj["removed"] as? [String] ?? []
        let added = obj["added"] as? [String] ?? []
        if !removed.isEmpty { try await client.put("/me/tracks", ids: removed) }
        if !added.isEmpty { try await client.delete("/me/tracks", ids: added) }
        return (removed.count, added.count)
    }

    private func writeLog(removed: [String], added: [String]) throws -> URL {
        let stampFormatter = DateFormatter()
        stampFormatter.dateFormat = "yyyyMMdd-HHmmss"
        let url = logDir.appendingPathComponent("apply-\(stampFormatter.string(from: Date())).json")
        let payload: [String: Any] = [
            "applied_at": ISO8601DateFormatter().string(from: Date()),
            "removed": removed,
            "added": added
        ]
        try JSONSerialization.data(withJSONObject: payload, options: [.prettyPrinted]).write(to: url)
        return url
    }
}

File: lib/spotify_sanitizer/apply.rb
# frozen_string_literal: true

require "json"
require "time"
require "fileutils"

module SpotifySanitizer
  # Executes a reviewed plan against the Spotify library, and writes a reversal
  # log so any run can be undone.
  #
  # Reversal semantics: unliking a track is reversible (re-like it); re-liking
  # is reversible (unlike it). The log records exactly what we changed.
  class Apply
    def initialize(client = Client.new, log_dir: File.join(Config.home, "logs"))
      @client = client
      @log_dir = log_dir
      FileUtils.mkdir_p(@log_dir)
    end

    def run(plan_hash, dry_run: false)
      remove_ids = Array(plan_hash["removals"]).map { |r| r["id"] }.compact
      add_ids    = Array(plan_hash["additions"]).map { |a| a["id"] }.compact

      if dry_run
        puts "DRY RUN — would unlike #{remove_ids.size}, like #{add_ids.size}. Nothing changed."
        return
      end

      @client.delete("/me/tracks", remove_ids) unless remove_ids.empty?
      @client.put("/me/tracks", add_ids)       unless add_ids.empty?

      log = write_log(remove_ids, add_ids)
      puts "Applied: unliked #{remove_ids.size}, liked #{add_ids.size}."
      puts "Reversal log: #{log}"
      puts "Undo with: spotify-sanitizer undo #{log}"
    end

    # Inverts a reversal log: re-like what we removed, unlike what we added.
    def undo(log_path)
      log = JSON.parse(File.read(log_path))
      @client.put("/me/tracks", log["removed"]) unless log["removed"].to_a.empty?
      @client.delete("/me/tracks", log["added"]) unless log["added"].to_a.empty?
      puts "Undone: re-liked #{log["removed"].to_a.size}, unliked #{log["added"].to_a.size}."
    end

    private

    def write_log(removed, added)
      path = File.join(@log_dir, "apply-#{Time.now.strftime("%Y%m%d-%H%M%S")}.json")
      File.write(path, JSON.pretty_generate(
        "applied_at" => Time.now.iso8601,
        "removed"    => removed,
        "added"      => added
      ))
      path
    end
  end
end
